src/connection.ts

```diff
@@ -967,42 +967,174 @@ export class Connection extends EventEmitter {
+    const prepare = {
+      destination: this.destinationAccount!,
+      amount: amount.toString(),
+      data: requestPacket.serializeAndEncrypt(this.sharedSecret),
+      executionCondition: cryptoHelper.generateRandomCondition(),
+      expiresAt: new Date(Date.now() + timeout)
+    }
 
-    const responsePacket = await this.sendPacket(requestPacket, sourceAmount, true)
-    if (!responsePacket) {
-      return
+    /* tslint:disable-next-line:no-unnecessary-type-assertion */
+    const responseData = await (new Promise(async (resolve, reject) => {
+      const timer = setTimeout(() => {
+        this.log.error(`test packet ${requestPacket.sequence} timed out before we got a response`)
+        resolve(null)
+      }, timeout)
+      const result = await this.plugin.sendData(IlpPacket.serializeIlpPrepare(prepare))
+      clearTimeout(timer)
+      resolve(result)
+    }) as Promise<Buffer | null>)
+
+    if (!responseData) {
+      return null
     }
 
-    this.remoteKnowsOurAccount = true
+    const ilpReject = IlpPacket.deserializeIlpReject(responseData)
+
+    // Return the receiver's response if there was one
+    let responsePacket
+    if (ilpReject.code === 'F99' && ilpReject.data.length > 0) {
+      responsePacket = Packet.decryptAndDeserialize(this.sharedSecret, ilpReject.data)
 
-    // Determine exchange rate from amount that arrived
-    this.exchangeRate = responsePacket.prepareAmount.dividedBy(sourceAmount)
-    this.log.debug(`determined exchange rate to be: ${this.exchangeRate}`)
-    if (this.exchangeRate.isEqualTo(0)) {
-      // TODO this could also happen if the exchange rate is less than 1 / TEST_PACKET_AMOUNT
-      throw new Error('Exchange rate is 0. We will not be able to send anything through this path')
+      // Ensure the response corresponds to the request
+      if (!responsePacket.sequence.isEqualTo(requestPacket.sequence)) {
+        this.log.error(`response packet sequence does not match the request packet. expected sequence: ${requestPacket.sequence}, got response packet:`, JSON.stringify(responsePacket))
+        throw new Error(`Response packet sequence does not correspond to the request. Actual: ${responsePacket.sequence}, expected: ${requestPacket.sequence}`)
+      }
+      if (responsePacket.ilpPacketType !== responseData[0]) {
+        this.log.error(`response packet was on wrong ILP packet type. expected ILP packet type: ${responseData[0]}, got:`, JSON.stringify(responsePacket))
+        throw new Error(`Response says it should be on an ILP packet of type: ${responsePacket.ilpPacketType} but it was carried on an ILP packet of type: ${responseData[0]}`)
+      }
+    } else {
+      this.log.debug(`test packet ${requestPacket.sequence} was rejected with a ${ilpReject.code} error${ilpReject.message ? ' with the message: "' + ilpReject.message + '"' : ''}`)
+    }
+
+    if (ilpReject.code[0] === 'T') {
```

Comment:
I like this logic, so basically it will keep retrying Txx errors for 30 seconds backing off more and more each time.